drivers/timer/nrf_rtc_timer.c

```diff
@@ -218,69 +214,67 @@ uint64_t z_nrf_rtc_timer_get_ticks(k_timeout_t t)
+	 * is checked).
+	 * Clear also the event as it may already be generated at this point.
+	 */
+	event_disable(chan);
+	event_clear(chan);
 
-	do {
-		now = counter();
+	for (;;) {
+		uint32_t now;
 
-		/* Handle case when previous event may generate an event.
-		 * It is handled by setting CC to now (far in the future),
-		 * in case previous event was set for next tick wait for half
-		 * LF tick and clear event that may have been generated.
+		set_comparator(chan, cc_val);
+		/* Enable event routing after the required CC value was set.
+		 * Even though the above operation may get repeated (see below),
+		 * there is no need to disable event routing in every iteration
+		 * of the loop, as the COMPARE event resulting from any attempt
+		 * of setting the CC register is acceptable (as mentioned above,
+		 * process_channel() does the proper filtering).
 		 */
-		set_comparator(chan, now);
-		if (counter_sub(prev_cc, now) == 1) {
-			/* It should wait for half of RTC tick 15.26us. As
-			 * busy wait runs from different clock source thus
-			 * wait longer to cover for discrepancy.
-			 */
-			k_busy_wait(19);
-		}
+		event_enable(chan);
 
-		/* RTC may not generate event if CC is set for 1 tick from now.
-		 * Because of that if requested cc_val is in the past or next tick,
-		 * set CC to further in future. Start with 2 ticks from now but
-		 * if that fails go even futher. It may fail if operation got
-		 * interrupted and RTC counter progressed or if optimization is
-		 * turned off.
-		 */
-		if (counter_sub(cc_val, now + 2) > COUNTER_HALF_SPAN) {
-			cc_val = now + tick_inc;
-			tick_inc++;
-		}
+		now = counter();
 
-		event_clear(chan);
-		event_enable(chan);
-		set_comparator(chan, cc_val);
-		now2 = counter();
-		prev_cc = cc_val;
-		/* Rerun the algorithm if counter progressed during execution
-		 * and cc_val is in the past or one tick from now. In such
-		 * scenario, it is possible that event will not be generated.
-		 * Rerunning the algorithm will delay the alarm but ensure that
-		 * event will be generated at the moment indicated by value in
-		 * CC register.
+		/* RTC may not generate a COMPARE event if its COUNTER value
+		 * is N and a given CC register is set to N or N+1. If it turns
+		 * out that the above configuration of the comparator resulted
+		 * in such CC value or even in a value that is considered to be
+		 * from the past, repeat the operation using a CC value that is
+		 * guaranteed to generate the event. Start with 2 RTC ticks from
+		 * now and if that fails (because the operation gets delayed),
+		 * go even futher in the next attempt.
+		 * But if the COMPARE event turns out to be already generated,
+		 * there is obviously no need to continue the loop.
 		 */
-	} while ((now2 != now) &&
-		 (counter_sub(cc_val, now2 + 2) > COUNTER_HALF_SPAN));
-
-	return cc_val;
+		if (counter_sub(cc_val, now + 2) > (COUNTER_HALF_SPAN - 2) &&
+		    !nrf_rtc_event_check(RTC, RTC_CHANNEL_EVENT_ADDR(chan))) {
```

Comment:
could you add wrapper `event_check` like in case of  `event_clear`?

Reply:
Added.